spec/language/pattern_matching_spec.rb

```diff
@@ -722,6 +725,8 @@ def obj.deconstruct; "" end
       end
 
       it "accepts a subclass of Array from #deconstruct" do
+        next skip if RUBY_ENGINE == "truffleruby"
+
```

Comment:
BTW this test case differs a bit from the one in the `ruby/spec` repository. Not sure if it's done intentionally. Here the test relies on a way how CRuby accesses array elements - with `#[]` method. The `ruby/spec`'s current version doesn't redefine the `#[]` method:

```ruby
    it "accepts a subclass of Array from #deconstruct" do
      obj = Object.new
      def obj.deconstruct
        Class.new(Array).new([0, 1])
      end

      eval(<<~RUBY).should == true
        case obj
        in [1, 2]
          false
        in [0, 1]
          true
        end
      RUBY
    end
```

This version passes on TruffleRuby but the version with overridden `#[]` - doesn't.

Reply:
Looks like the change to `ruby/spec` was made for TruffleRuby: https://github.com/ruby/spec/commit/0e042687332c12a9f0468b52c1dc12d4a5a7e0fa

Unfortunately, there are no additional details on that change.

MRI uses `#[]` (or `opt_aref`) to access elements returned by `#deconstruct`. So, you can do smth like this:

```ruby
class A < Array
  def [](i) = at(-i-1)
end

a = A.new([1, 2])

case a
in [2, 1]
  true
end #=> true
```

I don't know if it's an intended feature or not; @k-tsj could you please take a look?

Reply:
@palkan 
Yes, it's intentional. If `#[]` is redefined, it should be called,.

Reply:
@k-tsj Thank you!

@eregon ^ That's what I mentioned yesterday.

Reply:
I see. I'm not sure it's a good idea to have that behavior in Ruby, maybe something to discuss on ruby-core.
Subclassing Array and redefining `[]` on its own is already clearly a bad idea, so it seems a niche edge case no one should use.